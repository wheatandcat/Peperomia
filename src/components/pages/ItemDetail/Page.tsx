import React, { FC, memo } from 'react';
import {
  View,
  Text,
  Linking,
  TouchableOpacity,
  Dimensions,
  StyleSheet,
  useWindowDimensions,
} from 'react-native';
import Toast from 'react-native-root-toast';
import EStyleSheet from 'react-native-extended-stylesheet';
import { ItemDetailQuery } from 'queries/api/index';
import Header from 'components/molecules/ScheduleHeader/Header';
import Label from 'components/molecules/ScheduleDetail/Label';
import ItemDetailWrap from 'components/organisms/ItemWrap/ItemDetailWrap';
import theme from 'config/theme';

type Props = {
  date: string;
  itemDetail: ItemDetailQuery['itemDetail'];
  onDismiss: () => void;
};

export type ScheduleDetailType = {
  onOpenActionSheet: () => void;
};

const handleClick = (url: string) => {
  Linking.canOpenURL(url).then((supported) => {
    if (supported) {
      Linking.openURL(url);
    } else {
      const { height } = Dimensions.get('window');

      const toast = Toast.show('無効なリンクです', {
        duration: Toast.durations.LONG,
        position: height - 150,
        shadow: true,
        animation: true,
        hideOnPress: true,
        textColor: theme().color.red,
        delay: 0,
      });

      setTimeout(function () {
        Toast.hide(toast);
      }, 3000);
    }
  });
};

const Page: FC<Props> = (props) => {
  const windowHeight = useWindowDimensions().height;

  return (
    <ItemDetailWrap
      date={props.date}
      title={props.itemDetail?.title || ''}
      kind={props.itemDetail?.kind || ''}
      rightIcon="more-horiz"
      onCloseKeyBoard={() => null}
      onRightPress={() => null}
      onDismiss={props.onDismiss}
    >
      <View style={[estyles.root, { minHeight: windowHeight }]}>
        <Header kind={props.itemDetail?.kind || ''}>
          <Text numberOfLines={1} style={styles.title}>
            {props.itemDetail?.title}
          </Text>
        </Header>

        <View>
          {Boolean(props.itemDetail?.place) && (
            <View style={styles.container}>
              <Label text="集合場所" icon="map-marker-outline" width={95} />

              <View style={styles.memoContainer}>
                <Text style={estyles.memoText}>{props.itemDetail?.place}</Text>
              </View>
            </View>
          )}

          {Boolean(props.itemDetail?.url) && (
            <View style={styles.container}>
              <Label text="URL" icon="link" width={70} />

              <View style={styles.memoContainer}>
                <TouchableOpacity
                  onPress={() => handleClick(props.itemDetail?.url || '')}
                >
                  <Text
                    style={[estyles.memoText, { color: theme().color.sky }]}
                    numberOfLines={1}
                  >
                    {props.itemDetail?.url}
                  </Text>
                </TouchableOpacity>
              </View>
            </View>
          )}

          {Boolean(props.itemDetail?.memo) && (
            <View style={styles.container}>
              <Label text="メモ" icon="file-document-box-outline" width={70} />

              <View style={styles.memoContainer}>
                <Text style={estyles.memoText}>{props.itemDetail?.memo}</Text>
              </View>
            </View>
          )}
        </View>
      </View>
      <View style={estyles.footer} />
    </ItemDetailWrap>
  );
};

export default memo(Page);

const estyles = EStyleSheet.create({
  root: {
    backgroundColor: '$background',
    flex: 1,
    width: '100%',
    height: '100%',
  },
  timeText: {
    fontSize: 18,
    color: '$text',
    paddingHorizontal: theme().space(3),
  },
  memoText: {
    fontSize: 16,
    lineHeight: 24,
    color: '$text',
  },
  footer: {
    width: '100%',
    height: '150%',
    backgroundColor: '$background',
  },
});

const styles = StyleSheet.create({
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    color: theme().color.darkGray,
  },
  memoContainer: {
    paddingTop: theme().space(2),
    paddingBottom: theme().space(3),
    paddingHorizontal: theme().space(1),
  },
  container: {
    paddingHorizontal: theme().space(3),
    paddingTop: theme().space(2),
  },
});
