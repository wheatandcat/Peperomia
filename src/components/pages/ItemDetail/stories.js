import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { KIND_PARK } from 'peperomia-util';
import Page from './Page';

const data = {
  date: '2020-01-01 00:00:00',
  itemDetail: {
    id: '1',
    title: '新宿駅',
    kind: KIND_PARK,
    memo: 'メモ',
    moveMinutes: 0,
    url: 'https://peperomia.app/',
    place: 'place',
    priority: 1,
  },
};

storiesOf('pages', module).add('itemDetail', () => (
  <Page itemDetail={data.itemDetail} date={data.date} />
));
