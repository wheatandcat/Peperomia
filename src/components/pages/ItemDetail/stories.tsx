import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { KIND_PARK } from 'peperomia-util';
import { mockFn, StackNavigator } from 'storyBookUtils';
import Page from './Page';

const data = {
  date: '2020-01-01 00:00:00',
  itemDetail: {
    id: '1',
    title: '新宿駅',
    kind: KIND_PARK,
    memo: 'メモ',
    url: 'https://peperomia.app/',
    place: 'place',
    priority: 1,
  },
};

const Screen = () => {
  return (
    <Page
      itemDetail={data.itemDetail}
      date={data.date}
      onDismiss={mockFn('onDismiss')}
      onUpdate={mockFn('onUpdate')}
      onDelete={mockFn('onDelete')}
      onUpdateMain={mockFn('onUpdateMain')}
    />
  );
};

storiesOf('pages', module).add('itemDetail', () => (
  <StackNavigator screen={Screen} />
));
