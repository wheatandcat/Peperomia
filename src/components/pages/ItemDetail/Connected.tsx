import React, { memo, useCallback } from 'react';
import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import 'dayjs/locale/ja';
import { useItemDetailQuery } from 'queries/api/index';
import { Props as IndexProps } from './';
import Plain from './Plain';

dayjs.extend(advancedFormat);

type Props = IndexProps & {
  date: string;
  itemId: string;
  itemDetailId: string;
};

export type ConnectedType = {
  date: string;
  onDismiss: () => void;
  onUpdate: () => void;
};

const Connected: React.FC<Props> = (props) => {
  const { data, loading, error, refetch } = useItemDetailQuery({
    variables: {
      date: props.date,
      itemId: props.itemId,
      itemDetailId: props.itemDetailId,
    },
  });

  const onCallback = useCallback(async () => {
    await refetch();
  }, [refetch]);

  const onUpdate = useCallback(() => {
    props.navigation.navigate('EditItemDetail', {
      date: props.date,
      itemId: props.itemId,
      itemDetailId: props.itemDetailId,
      onCallback,
    });
  }, [props, onCallback]);

  const onDismiss = useCallback(() => {
    props.navigation.goBack();
  }, [props.navigation]);

  return (
    <Plain
      data={data}
      loading={loading}
      error={error}
      date={props.date}
      onDismiss={onDismiss}
      onUpdate={onUpdate}
    />
  );
};

export default memo(Connected);
