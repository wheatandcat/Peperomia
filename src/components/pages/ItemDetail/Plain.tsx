import alertDialog from 'lib/alertDialog';
import ErrorPage from 'components/organisms/Error/Error';
import React, { memo } from 'react';
import { ItemDetailQueryHookResult } from 'queries/api/index';
import Loading from 'components/atoms/Loading';
import { ConnectedType } from './Connected';
import Page from './Page';

export type QueryProps = Pick<
  ItemDetailQueryHookResult,
  'data' | 'loading' | 'error'
>;

type Props = QueryProps & ConnectedType;

const Plain: React.FC<Props> = (props) => {
  if (alertDialog(props)) return <ErrorPage {...props} />;
  if (props.loading) return <Loading loading />;

  const itemDetail = props.data?.itemDetail!;

  return (
    <Page
      itemDetail={itemDetail}
      date={props.date}
      onDismiss={props.onDismiss}
    />
  );
};

export default memo(Plain);
