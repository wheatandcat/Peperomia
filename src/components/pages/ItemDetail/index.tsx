import React, { memo } from 'react';
import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import { RootStackParamList } from 'lib/navigation';
import Connected from './Connected';

type ScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'ItemDetail'
>;
type ScreenRouteProp = RouteProp<RootStackParamList, 'ItemDetail'>;

export type Props = {
  navigation: ScreenNavigationProp;
  route: ScreenRouteProp;
};

const itemDetail: React.FC<Props> = (props) => {
  const date = props.route.params.date;
  const itemId = props.route.params.itemId;
  const itemDetailId = props.route.params.itemDetailId;
  const onCallback = props.route.params.onCallback;

  return (
    <Connected
      {...props}
      date={date}
      itemId={itemId}
      itemDetailId={itemDetailId}
      onCallback={onCallback}
    />
  );
};

export default memo(itemDetail);
