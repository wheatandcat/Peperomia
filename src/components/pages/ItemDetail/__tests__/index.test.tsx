import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import ItemDetail, { Props } from '../';

const propsData = (): Props =>
  ({
    navigation: {
      setParams: jest.fn(),
      navigate: jest.fn(),
    },
    route: {
      params: {
        date: '2020-01-01 00:00:00',
        itemId: 'test',
        itemDetailId: 'test',
        onCallback: jest.fn(),
      },
    },
  } as any);

describe('components/pages/ItemDetail/index.tsx', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<ItemDetail {...propsData()} />);
  });

  it('正常にrenderすること', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
