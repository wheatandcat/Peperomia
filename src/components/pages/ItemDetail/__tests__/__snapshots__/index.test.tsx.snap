// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/ItemDetail/index.tsx 正常にrenderすること 1`] = `
<Memo(Connected)
  date="2020-01-01 00:00:00"
  itemDetailId="test"
  itemId="test"
  navigation={
    Object {
      "navigate": [MockFunction],
      "setParams": [MockFunction],
    }
  }
  onCallback={[MockFunction]}
  route={
    Object {
      "params": Object {
        "date": "2020-01-01 00:00:00",
        "itemDetailId": "test",
        "itemId": "test",
        "onCallback": [MockFunction],
      },
    }
  }
/>
`;
